### ISSUE TITLE
export_default_credentials: true isn't setting $GOOGLE_APPLICATION_CREDENTIALS for later steps

### ISSUE THREAD

**Issue by brachi-wernick** [GoogleCloudPlatform/github-actions#80]
### TL;DR
I set `export_default_credentials: true`, and in later steps test failed due to missing credentials.

**Expected behavior**
`$GOOGLE_APPLICATION_CREDENTIALS` should be set with the credentials for all later steps.

### Reproduction

**Action YAML**

```yaml
jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v1
    - uses: GoogleCloudPlatform/github-actions/setup-gcloud@master
      with:
       version: '270.0.0'
       service_account_email: ${{ secrets.GCP_SA_STG_EMAIL }}
       service_account_key: ${{ secrets.GCP_SA_STG_KEY }}
       export_default_credentials: true
    - name: set gcp cradentilas
      run:  gcloud info
    - name: echo gcp cradentilas
      run:  echo this is  ${GOOGLE_APPLICATION_CREDENTIALS}  
    - uses: actions/checkout@v2
    - name: Set up JDK 1.8
      uses: actions/setup-java@v1
      with:
        java-version: 1.8
    - name: Build with Maven
      run: mvn clean install --file pom.xml
```

**Additional information**
I submitted in the past some [PR](https://github.com/GoogleCloudPlatform/github-actions/pull/30) that fixed it, when I run same action with that PR branch it works.

**Comment by sethvargo** [comment]
Hi @brachi-wernick - thank you for opening an issue. The action does export `GOOGLE_APPLICATION_CREDENTIALS` for use in later steps [here](https://github.com/GoogleCloudPlatform/github-actions/blob/master/setup-gcloud/src/setup-gcloud.ts#L134).

Do you see the following line in your output for the `setup-gcloud` action?

```
Successfully exported Default Application Credentials!
```

If this is an open source project, can you link to the action run please?

**Comment by brachi-wernick** [comment]
I see in the logs ```Successfully exported Default Application Credentials```
I think the issue here that you variable exists in further steps but its value, the file itself , isn't available/ accessable
you should try export the origin path: `tmpKeyFilePath`
```core.exportVariable('GOOGLE_APPLICATION_CREDENTIALS', tmpKeyFilePath);```

**Comment by sethvargo** [comment]
**TL;DR: Update to actions/checkout@v2**

According to the [GitHub docs](https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#filesystems-on-github-hosted-runners), the `GITHUB_WORKSPACE` environment variable is shared among subsequent actions. Furthermore, our [integration test for this feature](https://github.com/GoogleCloudPlatform/github-actions/blob/c6a3086a37d48d7adeb1cd563f31c1fe2f9fce11/.github/workflows/setup-gcloud-it.yml#L76) explicitly tests that the file path defined by the environment variable `GOOGLE_APPLICATION_CREDENTIALS` exists and is a real file.

I noticed that you're using actions/checkout@v1. This is an old version that does not checkout code into the `GITHUB_WORKSPACE` path. GitHub recommends actions share data using only this workspace directory. I was able to get the expected behavior by updating your action to use v2 instead of v1. You can check the [successful run output](https://github.com/sethvargo/github-actions/runs/549021719?check_suite_focus=true). Note: it fails because I don't actually have any Java files or maven setup, but you can see the credentials are exported and the subsequent steps correctly echo the file path.

I believe your previous PR works because it uses a global `/tmp` path, but this is against GitHub's recommendations for sharing data, and could break in non-backwards-compatible ways in the future. I opened https://github.com/GoogleCloudPlatform/github-actions/pull/87 to document the requirement to use v2 of the checkout command.

I'm really sorry this didn't work as expected on the first try, but I think we've successfully debugged the issue and provided adequate steps to unblock you. I'm going to close out this issue, but if this isn't a good resolution, or if you have additional questions, please comment and we'll re-open it. Thank you!

**Comment by brachi-wernick** [comment]
@sethvargo thank you for your explanation, I copied your action flow, to make sure I didn't forget something. and still it failed in my tests, java application doesn't have the creds, even with v2.
how did you prove that in java you can access that env creds file? in your example there is no even pom.

**Comment by sethvargo** [comment]
Hi @brachi-wernick - I did not. As I said in my comment, I verified that GOOGLE_APPLICATION_CREDENTIALS is set to a file containing the credentials and that file exists on disk. I don't have access to your Java code to validate how you're reading the credential.

**Comment by brachi-wernick** [comment]
@sethvargo , I really see issue here. I try reproduce that, look at [build](https://github.com/brachi-wernick/testAction/runs/549684340?check_suite_focus=true) that failed in the maven build step. (after the build failed, I remove the secrets from github, since that is my company secrets).
I also try to commit it into your fork, in that [PR](https://github.com/sethvargo/github-actions/pull/3). (from some reason the check didn't run, do you know why?)

can you look and check why this can't find the creds?
even if you see that the env variable is available, that doesn't says that the file exists.

**Comment by sethvargo** [comment]
@brachi-wernick remove the second `actions/checkout@v2`. Performing a checkout resets the workspace, effectively deleting the cached credentials. Again, this works with your PR because the credentials are saved outside of the workspace.

- Commit: https://github.com/sethvargo/github-actions/commit/d53115e11595ae03ee49038f99ce7bc2cd2c22d2
- "Successful" run: https://github.com/sethvargo/github-actions/runs/550149962?check_suite_focus=true

Note: the run failed because I do not have access to your bucket. However, the authentication is clearly selected as the error message now includes the service account name.

Please let me know if you have additional questions.

**Comment by brachi-wernick** [comment]
Hi @sethvargo. that helps! it works now!
thanks on your help!

**Comment by ninabohm** [comment]
> @brachi-wernick remove the second `actions/checkout@v2`. Performing a checkout resets the workspace, effectively deleting the cached credentials. Again, this works with your PR because the credentials are saved outside of the workspace.
> 
> * Commit: https://github.com/sethvargo/github-actions/commit/d53115e11595ae03ee49038f99ce7bc2cd2c22d2
> * "Successful" run: https://github.com/sethvargo/github-actions/runs/550149962?check_suite_focus=true
> 
> Note: the run failed because I do not have access to your bucket. However, the authentication is clearly selected as the error message now includes the service account name.
> 
> Please let me know if you have additional questions.

your answer just saved my life @sethvargo ! ❤️ 
 I was debugging this error when trying to push to artifact registry `denied: Permission "artifactregistry.repositories.uploadArtifacts" denied on resource [...]` 
since several days and you just made me aware that the second checkout was the issue the entire time!
